Sources/Core/Microsoft.StreamProcessing/Operators/Clip/PartitionedClipJoinPipe.cs

```diff
@@ -365,9 +366,15 @@ private void ProcessPendingEntries()
                     var r = this.rightQueue.entries[index];
                     if (l.value.Count == 0 && r.value.Count == 0)
                     {
-                        this.seenKeys.Remove(p);
-                        this.leftQueue.Remove(p);
-                        this.rightQueue.Remove(p);
+                        this.partitionData.Lookup(p, out index);
+                        var partition = this.partitionData.entries[index].value;
+
+                        if (partition.IsClean())
```

Comment:
Why are we testing for "IsClean" in two different places?

Reply:
Why are we testing Count == 0 in two different places :)

---
In reply to: [285214635](https://github.com/microsoft/Trill/pull/89#discussion_r285214635) [](ancestors = 285214635)

Reply:
Not sure, but you're testing if a partition is clean before adding it to cleanKeys, and then testing it again inside the cleanKeys loop.